taraxa/state/dpos/precompiled/dpos_contract.go

```diff
@@ -1000,15 +1040,17 @@ func (self *Contract) redelegate(ctx vm.CallFrame, block types.BlockNum, args so
 		}
 
 		if validator_from.TotalStake.Cmp(big.NewInt(0)) == 0 && validator_rewards_from.CommissionRewardsPool.Cmp(big.NewInt(0)) == 0 {
-			self.validators.DeleteValidator(&args.ValidatorFrom)
-			self.state_put(&state_k, nil)
+			if !self.isMagnoliaHardfork(block) || validator_from.UndelegationsCount == 0 {
+				self.validators.DeleteValidator(&args.ValidatorFrom)
+				self.state_put(&state_k, nil)
```

Comment:
@MatusKysel same here, is it ok that I delete state only when I delete validator ?

Reply:
hmm and shouldnt I use self.state_get_and_decrement(args.Validator[:], BlockToBytes(validator.LastUpdated)) here ?